### PR TITLE
Document popular processing capabilities

### PR DESCRIPTION
**Pull request by dedemorton**
Adds doc enhancement requested in https://github.com/elastic/logstash/issues/6820

@acchen97 Note that this content can't be formatted in tables because the examples would cause text overruns in the formatted doc. The best compromise is definition lists. I've broken the content into separate topics, but we could just as easily have one topic. I think separate topics are easier to skim. 

@suyograo or @jordansissel Hoping that one of you can also review this. Our plugin docs are woefully lacking in examples and some of the descriptions are hard to follow. :-/ Where possible, I just reused examples from the docs, but in some cases, I had to create new examples. There are a couple of places where I need help coming up with a good example because I didn't have time to create it myself or felt the example in the existing doc was too complicated for an overview.

ALL: Please respond to questions that I've added as comments (addressed to either you or general reviewers).

Here's a screen grab that shows the basic info architecture of these topics:

image

### CODE REVIEW

File: docs/static/transforming-data.asciidoc [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
++
+The following config dissects messages that match the structure specified in
+the mapping:
++
+[source,json]
+--------------------------------------------------------------------------------
+filter {
+  dissect {
+    mapping => {
+      "message" => "%{ts} %{+ts} %{+ts} %{src} %{} %{prog}[%{pid}]: %{msg}"
```
Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: Let's use this example (which avoids the possibly confusing empty  `%{}` usage):

```
mapping => { "message" => "%{ts} %{+ts} %{+ts} %{src} %{prog}[%{pid}]: %{msg}" 
```

Sample input:

```
Apr 26 12:20:02 localhost systemd[1]: Starting system activity accounting tool...
```

Sample output (as through `stdout { codec => rubydebug }`:

```
{
  "msg"        => "Starting system activity accounting tool...",
  "@timestamp" => 2017-04-26T19:33:39.257Z,
  "src"        => "localhost",
  "@version"   => "1",
  "host"       => "localhost.localdomain",
  "pid"        => "1",
  "message"    => "Apr 26 12:20:02 localhost systemd[1]: Starting system activity accounting tool...",
  "type"       => "stdin",
  "prog"       => "systemd",
  "ts"         => "Apr 26 12:20:02"
}
```

File: docs/static/transforming-data.asciidoc [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+
+Copies fields from previous log events in Elasticsearch to current events. 
++
+[source,json]
+--------------------------------------------------
+NEED A SIMPLE EXAMPLE HERE. 
+--------------------------------------------------
+
+
+//REVEIWERS: I NEED A BETTER EXAMPLE FOR THE ELASTICSEARCH FILTER. THE EXAMPLE IN THE DOCUMENTATION MIGHT BE TOO COMPLEX FOR AN OVERVIEW.
```
Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: +1 for needing an example. I don't remember ever using this filter, but I will try to come up with an example.

File: docs/static/transforming-data.asciidoc [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+The following example executes a SQL query and stores the result set in a field
+called `country_details`:
++
+[source,json]
+--------------------------------------------------------------------------------
+filter {
+  jdbc_streaming {
+    jdbc_driver_library => "/path/to/mysql-connector-java-5.1.34-bin.jar"
+    jdbc_driver_class => "com.mysql.jdbc.Driver"
+    jdbc_connection_string => ""jdbc:mysql://localhost:3306/mydatabase"
```
Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: too many doublequotes before the value `=> "jdbc:mysql...`

File: docs/static/transforming-data.asciidoc [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+    parameters => { "code" => "country_code"}
+    target => "country_details"
+  }
+}
+--------------------------------------------------------------------------------
+
+
+//REVIEWERS: THE MYSQL CONNECTOR SHOWN IN THIS EXAMPLE IS KIND OF OLD. NOT SURE I SHOULD CHANGE IT, THOUGH, BECAUSE I'M NOT SURE IF WE SUPPORT ALL VERSIONS OF THE CONNECTOR.
+
+//ALVIN: YOU ALSO HAD JDBC STATIC PLUGIN IN YOUR LIST. DO YOU MEAN THE INPUT PLUGIN? SEEMS A LITTLE OUT-OF-PLACE TO LIST AN INPUT PLUGIN IN THIS SECTION, IMO. DOES IT DO SOMETHING BEYOND WHAT INPUT PLUGINS NORMALLY DO SO THAT IT
```
Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: There is a jdbc_static filter that is [in progress](https://github.com/logstash-plugins/logstash-filter-jdbc_static/pulls) but not yet released.

File: docs/static/transforming-data.asciidoc [hunk 5]
```diff
@@ -0,0 +1,520 @@
...
+    jdbc_password => "secret"
+    statement => "select * from WORLD.COUNTRY WHERE Code = :code"
+    parameters => { "code" => "country_code"}
+    target => "country_details"
+  }
+}
+--------------------------------------------------------------------------------
+
+
+//REVIEWERS: THE MYSQL CONNECTOR SHOWN IN THIS EXAMPLE IS KIND OF OLD. NOT SURE I SHOULD CHANGE IT, THOUGH, BECAUSE I'M NOT SURE IF WE SUPPORT ALL VERSIONS OF THE CONNECTOR.
```
Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: I don't know if we specify what connectors (or versions of a connector) are supported. @guyboertje maybe has ideas?

File: docs/static/transforming-data.asciidoc [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+
+<<plugins-filters-aggregate,aggregate filter>>::
+
+Aggregates information from several events (typically log lines) that belong to
+the same task and pushes the aggregated information into a final task event.
+For example:
++
+[source,json]
+--------------------------------------------------------------------------------
+filter {
```
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: @jordansissel Right. I understand your concern. I've sorted the options alphabetically just because I think it's easier for people to scan. This topic isn't a reference topic, though, so I would be OK with listing the plugins in order of usefulness. Should I use the order that Alvin gives here? https://github.com/elastic/logstash/issues/6820

File: docs/static/transforming-data.asciidoc [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+--------------------------------------------------------------------------------
+input {
+  file {
+    path => "/path/to/myfile.json"
+    codec =>"json"
+}
+--------------------------------------------------------------------------------
+
+
+//ALVIN: SHOULD WE INCLUDE JSON_LINES HERE, TOO, OR MAYBE MENTION IT?
```
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: @jordansissel So if someone uses a community plugin that's not smart enough to do this magic...then it just blows up? That seems...not so good. I wonder why json_lines isn't just part of the json codec?

If the json codec works most of the time, then I vote for not covering json_lines.

File: docs/static/transforming-data.asciidoc [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+
+<<plugins-codecs-protobuf,protobuf codec>>::
+
+Reads protobuf encoded messages and converts them to Logstash events. Requires
+the protobuf definitions to be compiled as Ruby files.
```
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: @jordansissel We cover that in the docs for the protobuf codec. Not sure it makes sense to go into that level of detail in this particular topic. Although..I gotta say...I sure like the idea of recommending something called "beefcake" in the doc.s :-P

File: docs/static/transforming-data.asciidoc [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+--------------------------------------------------------------------------------
+
+
+//REVIEWERS: PLEASE DOUBLE-CHECK THE SYNTAX FOR THE PROTOBUF EXAMPLE. I CHANGED THE FORMATTING TO BE CONSISTENT WITH THE REST OF THE DOCUMENTATION HERE, BUT DIDN'T TEST THIS EXAMPLE.
+
+<<plugins-filters-xml,xml filter>>::
+
+Parses XML into fields.
++
+The following config parses the whole XML document stored in the `message` field:
```
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: I'd like to get @acchen97 to weigh in on this one. I've tried to strike a balance between providing sufficient info, but not not too much info because we want users to have something they can read quickly. For some of the examples, I started adding the input message and result because I thought it was hard to understand the plugin without seeing that (especially for stuff like grok). In this case, I think it's worthwhile. But maybe Alvin should see if there are other places where input/output examples would be useful.

File: docs/static/transforming-data.asciidoc [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+The plugins described in this section are useful for extracting fields and
+parsing unstructured data into fields.
+
+Also see <<filter-plugins>> for the full list of available filter plugins. 
+
+<<plugins-filters-dissect,dissect filter>>::
+
+Extracts unstructured event data into fields by using delimiters. The dissect
+filter does not use regular expressions and is very fast. However, if the
+structure of the data varies from line to line, the grok filter is more
```
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: @jordansissel If you believe the doc. :-) What do you think?

File: docs/static/transforming-data.asciidoc [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+* <<field-extraction>>
+* <<lookup-enrichment>>
+
+[[core-operations]]
+=== Performing Core Operations
+
+The plugins described in this section are useful for core operations, such as
+aggregating, mutating, and dropping events.
+
+Also see <<filter-plugins>> for the full list of available filter plugins.
```
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: Instead of having this note in each section, should we just blanket it in the "Transforming Data" intro?

Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: Sure we can do that...the repetition is for users who jump around and don't read from the beginning to end. I don't have a strong preference.

File: docs/static/transforming-data.asciidoc (hunk 6 above)
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: +1 on the using ordering in that issue. I agree that aggregate shouldn't be the first item.

Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: I'm in favor of not mentioning the aggregate filter here at all. We are working on finding a better solution that could replace the aggregate filter; no details at this time though.

File: docs/static/transforming-data.asciidoc [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+
+Parses dates from fields to use as Logstash timestamps for events.
++
+The following config parses a field called `logdate` to set the Logstash
+timestamp:
++
+[source,json]
+--------------------------------------------------------------------------------
+filter {
+  date {
```
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: I feel this one is relatively self-explanatory and fine as-is. If they want more details, they can drill into the respective plugin page.

File: docs/static/transforming-data.asciidoc (hunk 7 above)
Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: @dedemorton I think the only two plugins that need "smarts" are the tcp input and stdin input. Any other input is likely never going to use the json_lines codec (and would use `json` instead where json is needed)

File: docs/static/transforming-data.asciidoc [hunk 13]
```diff
@@ -0,0 +1,520 @@
...
+Drops events. This filter is typically used in combination with conditionals.
++
+The following config drops 40% of the `debug` level log messages:
++
+[source,json]
+--------------------------------------------------------------------------------
+filter {
+  if [loglevel] == "debug" {
+    drop {
+      percentage => 40
```
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: In this context, I think just showing the most popular case (#1) is sufficient.

File: docs/static/transforming-data.asciidoc (hunk 9 above)
Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: > For some of the examples, I started adding the input message and result because I thought it was hard to understand 

I agree with your reasoning. For XML I agree the input may not need to be shown.

File: docs/static/transforming-data.asciidoc (hunk 10 above)
Comment (jordansissel) [inline docs/static/transforming-data.asciidoc]: haha. I think it's true. I had some misunderstandings about dissect's behavior last week and I think I understand it more now.

**Comment by jordansissel** [comment]
Did another round commenting.

File: docs/static/transforming-data.asciidoc [hunk 14]
```diff
@@ -0,0 +1,520 @@
...
+Performs general mutations on fields. You can rename, remove, replace, and
+modify fields in your events.
++
+The following config renames the `HOSTORIP` field to `client_ip`:
++
+[source,json]
+--------------------------------------------------------------------------------
+filter {
+  mutate {
+    rename => { "HOSTORIP" => "client_ip" }
```
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: I'd probably add a few other usages into the mutate block here as they're relatively simple. Perhaps add 'strip' and 'remove_field' examples?

File: docs/static/transforming-data.asciidoc [hunk 15]
```diff
@@ -0,0 +1,520 @@
...
+--------------------------------------------------------------------------------
+
+
+[[data-deserialization]]
+=== Deserializing Data
+
+The plugins described in this section are useful for deserializing data into
+Logstash events.
+
+Also see <<filter-plugins>> and <<codec-plugins>> for the full list of available
```
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: Same for this comment as the one above. Maybe just note the full list of filters and codecs are available.

File: docs/static/transforming-data.asciidoc [hunk 16]
```diff
@@ -0,0 +1,520 @@
...
+input {
+  tcp {
+    codec => fluent
+    port => 4000
+  }
+}
+--------------------------------------------------------------------------------
+
+
+//ALVIN: SHOULD WE ALSO COVER THE MSGPACK CODEC? I LOOKED AT THE DOCS FOR MSGPACK AND THERE'S NO USEFUL INFO IN THERE.
```
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: I'm ok not including it.

File: docs/static/transforming-data.asciidoc (hunk 7 above)
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: In that case, I'm +1 on not including json_lines.

@dedemorton it sounds like it doesn't blow up, it'll handle the conversion to json codec under the covers if I'm understanding correctly.

File: docs/static/transforming-data.asciidoc [hunk 17]
```diff
@@ -0,0 +1,520 @@
...
+  topic_id => "your_topic_goes_here"
+  codec => protobuf {
+    class_name => "Animal::Unicorn"
+    include_path => ['/path/to/protobuf/definitions/UnicornProtobuf.pb.rb']
+  }
+}
+--------------------------------------------------------------------------------
+
+
+//REVIEWERS: PLEASE DOUBLE-CHECK THE SYNTAX FOR THE PROTOBUF EXAMPLE. I CHANGED THE FORMATTING TO BE CONSISTENT WITH THE REST OF THE DOCUMENTATION HERE, BUT DIDN'T TEST THIS EXAMPLE.
```
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: @IngaFeick if you don't mind, can you help confirm the Protobuf codec syntax here? We're looking to highlight it better in our docs.

File: docs/static/transforming-data.asciidoc (hunk 9 above)
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: I think its fine as-is, this is a feature discovery page.

File: docs/static/transforming-data.asciidoc (hunk 1 above)
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: ++ I like the example and the sample input/output is valuable esp in the field extraction context.

File: docs/static/transforming-data.asciidoc (hunk 2 above)
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: Pinging @gingerwizard who has experience using this filter. Perhaps he has some ideas for a simple example here.

File: docs/static/transforming-data.asciidoc (hunk 4 above)
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: Yes, it's in progress and getting pretty close. We can add it later after it gets in formally.

File: docs/static/transforming-data.asciidoc (hunk 6 above)
Comment (acchen97) [inline docs/static/transforming-data.asciidoc]: I'm ok with omitting it and its more of an advanced rather than core operation. There's already a myriad of good content here.

File: docs/static/transforming-data.asciidoc (hunk 8 above)
Comment (IngaFeick) [inline docs/static/transforming-data.asciidoc]: @jordansissel I'm using the ruby-protoc by codekitchen. I put that in the README but I guess it makes sense to add it here aswell.

File: docs/static/transforming-data.asciidoc (hunk 17 above)
Comment (IngaFeick) [inline docs/static/transforming-data.asciidoc]: @acchen97 lgtm

File: docs/static/transforming-data.asciidoc (hunk 11 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: done

File: docs/static/transforming-data.asciidoc (hunk 6 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: done - Removed aggregate. I'm out of time for reordering the content. If we think that's necessary even after removing aggregate, I can do it in a separate PR later.

File: docs/static/transforming-data.asciidoc (hunk 12 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: agreed...leaving as-is

File: docs/static/transforming-data.asciidoc (hunk 13 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: fixed to show only the first use case.

File: docs/static/transforming-data.asciidoc (hunk 14 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: fixed - I've added strip, but not remove_field because that's actually one of the common filter options (I know...it's hard to tell...good thing we're fixing that)

File: docs/static/transforming-data.asciidoc (hunk 15 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: done

File: docs/static/transforming-data.asciidoc (hunk 16 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: left as-is

File: docs/static/transforming-data.asciidoc (hunk 7 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: left as-is

File: docs/static/transforming-data.asciidoc (hunk 8 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: added link to ruby-protoc compiler (people will need to find beefcake on their own...I can't be implicated in that).

File: docs/static/transforming-data.asciidoc [hunk 18]
```diff
@@ -0,0 +1,520 @@
...
+<<plugins-codecs-protobuf,protobuf codec>>::
+
+Reads protobuf encoded messages and converts them to Logstash events. Requires
+the protobuf definitions to be compiled as Ruby files.
++
+The following config decodes events from a Kafka stream:
++
+[source,json]
+--------------------------------------------------------------------------------
+kafka {
```
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: fixed

File: docs/static/transforming-data.asciidoc (hunk 17 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: left as-is

File: docs/static/transforming-data.asciidoc (hunk 9 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: left as-is

File: docs/static/transforming-data.asciidoc (hunk 10 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: left as-is

File: docs/static/transforming-data.asciidoc (hunk 1 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: added

File: docs/static/transforming-data.asciidoc (hunk 2 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: @gingerwizard @jordansissel I'm hoping to wrap this content up before I go on vacation. Does either of you have a suggested for an example that is a bit less complex than what we show in the doc for the ES filter?

File: docs/static/transforming-data.asciidoc (hunk 3 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: fixed

File: docs/static/transforming-data.asciidoc (hunk 5 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: minor issue, so leaving as-is. I can update in a separate PR when guy responds.

File: docs/static/transforming-data.asciidoc (hunk 4 above)
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: leaving as-is. Just let me know when it's ready to be added to the list.

File: docs/static/transforming-data.asciidoc [hunk 19]
```diff
@@ -0,0 +1,532 @@
...
+            hosts => ["es-server"]
+            query => "type:start AND operation:%{[opid]}"
+            fields => { "@timestamp" => "started" }
+         }
+         date {
+            match => ["[started]", "ISO8601"]
+            target => "[started]"
+         }
+         ruby {
+            code => "event['duration_hrs'] = (event['@timestamp'] - event['started']) / 3600 rescue nil"
```
Comment (dedemorton) [inline docs/static/transforming-data.asciidoc]: With the new event API syntax, I think this should actually be:

```
code => "event.set("[duration_hrs]") = (event.get("[@timestamp]") - event.get("[started]")) / 3600 rescue nil"
```

Comment (suyograo) [inline docs/static/transforming-data.asciidoc]: `code => "event.set("duration_hrs", (event.get("@timestamp") - event.get("started")) / 3600) rescue nil"`

Comment (suyograo) [inline docs/static/transforming-data.asciidoc]: @dedemorton use this syntax ^^

**Comment by jordansissel** [comment]
🤘

**Comment by acchen97** [comment]
@dedemorton this is really fantastic work! I just checked it out live and users will definitely feel an impact on the getting started experience.
